libsolidity/analysis/ConstantEvaluator.cpp

```diff
@@ -32,85 +32,367 @@ using namespace solidity;
 using namespace solidity::frontend;
 using namespace solidity::langutil;
 
-void ConstantEvaluator::endVisit(UnaryOperation const& _operation)
+using TypedRational = ConstantEvaluator::TypedRational;
+
+namespace
+{
+
+/// Check whether (_base ** _exp) fits into 4096 bits.
+bool fitsPrecisionExp(bigint const& _base, bigint const& _exp)
 {
-	auto sub = type(_operation.subExpression());
-	if (sub)
-		setType(_operation, sub->unaryOperatorResult(_operation.getOperator()));
+	if (_base == 0)
+		return true;
+
+	solAssert(_base > 0, "");
+
+	size_t const bitsMax = 4096;
+
+	unsigned mostSignificantBaseBit = boost::multiprecision::msb(_base);
+	if (mostSignificantBaseBit == 0) // _base == 1
+		return true;
+	if (mostSignificantBaseBit > bitsMax) // _base >= 2 ^ 4096
+		return false;
+
+	bigint bitsNeeded = _exp * (mostSignificantBaseBit + 1);
+
+	return bitsNeeded <= bitsMax;
 }
 
-void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
+/// Checks whether _mantissa * (2 ** _expBase10) fits into 4096 bits.
+bool fitsPrecisionBase2(bigint const& _mantissa, uint32_t _expBase2)
```

Comment:
Why was `fitsPrecisionBaseX` moved to `Common` but not these 2?

Reply:
It is not needed there.